### PR TITLE
aliases

### PR DESCRIPTION
**Pull request by izulin**
### Context

New mechanizm of declaring function aliases inside function plugins.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1.
2.
3.

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation,
- [x] I described the modification in the CHANGELOG.md file.

### CODE REVIEW

**Comment by izulin** [comment]
@wojciechczerniak runs are failing at tests from nonexisting files

**Comment by wojciechczerniak** [comment]
> @wojciechczerniak runs are failing at tests from nonexisting files

@izulin No, the file exists https://github.com/handsontable/hyperformula/blob/pu/aliases/test/interpreter/function-negbinomdist.spec.ts and this one https://github.com/handsontable/hyperformula/blob/pu/aliases/test/interpreter/function-normsinv.spec.ts and they fail

**Comment by izulin** [comment]
> > @wojciechczerniak runs are failing at tests from nonexisting files
> 
> @izulin Are you sure? This file exists https://github.com/handsontable/hyperformula/blob/pu/aliases/test/interpreter/function-negbinomdist.spec.ts and this one https://github.com/handsontable/hyperformula/blob/pu/aliases/test/interpreter/function-normsinv.spec.ts and they fail

Right. I thought I was going crazy 💯

File: src/interpreter/FunctionRegistry.ts [hunk 1]
```diff
@@ -7,13 +7,37 @@ import {FunctionMetadata, FunctionPlugin, FunctionPluginDefinition} from './plug
 import {Interpreter} from './Interpreter'
 import {Maybe} from '../Maybe'
 import {Config} from '../Config'
-import {AliasAlreadyExisting, AliasTargetEmpty, FunctionPluginValidationError, ProtectedFunctionError} from '../errors'
+import {
+  AliasAlreadyExisting,
+  AliasEmpty,
+  AliasTargetEmpty,
```
Comment (voodoo11) [inline src/interpreter/FunctionRegistry.ts:13]: Unused

Comment (izulin) [reply src/interpreter/FunctionRegistry.ts:13]: :+1:

**Review by voodoo11 — APPROVED** [review]
👍

File: src/HyperFormula.ts [hunk 2]
```diff
@@ -423,7 +423,7 @@ export class HyperFormula implements TypedEmitter {
   }
 
   /**
-   * Registers a function with a given id if such exists in a plugin
+   * Registers a function with a given id if such exists in a plugin (as a function or an alias).
```
Comment (wojciechczerniak) [inline src/HyperFormula.ts:426]: I'm not sure how to do it. An example would be nice. We can throw a new error here if I understood the code correctly. And the description is now harder to understand. "Registers a function .. as a function". Should we rephrase the whole description? CC @scarletfog

File: docs/guide/custom-functions.md [hunk 3]
```diff
@@ -57,7 +57,33 @@ class CountHF extends FunctionPlugin {
...
+    // this method's functionality will be defined in the next step
+      method: 'hyper',
+    }
+  };
+  public static aliases = {
+    'HYPER.ALIAS': 'HYPER'
+  //HYPER.ALIAS is now an alias to HYPER
+  };
+}
+```
```
Comment (wojciechczerniak) [inline docs/guide/custom-functions.md:86]: @scarletfog New docs, adding you as a reviewer

File: src/HyperFormula.ts (hunk 2 above)
Comment (wojciechczerniak) [reply src/HyperFormula.ts:426]: I think that `registerFunctionPlugin` method has changed internally as well. Not sure how do we handle `unregister*` methods when it comes to aliases @izulin?

File: src/interpreter/FunctionRegistry.ts [hunk 4]
```diff
@@ -30,12 +49,8 @@ export class FunctionRegistry {
...
 
   public static registerFunction(functionId: string, plugin: FunctionPluginDefinition, translations?: FunctionTranslationsPackage): void {
-    const entry = plugin.implementedFunctions[functionId]
-    if (entry !== undefined) {
-      this.loadPluginFunction(plugin, functionId, this.plugins)
-    } else {
-      throw FunctionPluginValidationError.functionNotDeclaredInPlugin(functionId, plugin.name)
-    }
+    validateAndReturnMetadataFromName(functionId, plugin)
+    this.loadPluginFunction(plugin, functionId, this.plugins)
```
Comment (wojciechczerniak) [inline src/interpreter/FunctionRegistry.ts:53]: Is `validateAndReturnMetadataFromName` necessary here? We will call it again in `loadPluginFunction` -> `loadFunctionUnprotected` first line. Calling those checks twice looks like an unnecessary overhead

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]

File: src/HyperFormula.ts (hunk 2 above)
Comment (izulin) [reply src/HyperFormula.ts:426]: Aliases are registered just as functions, so we are able to unregister them exactly in the same way.

Comment (izulin) [reply src/HyperFormula.ts]: Ok, I agree that the explanation is a little confusing. I reverted it back.

File: docs/guide/custom-functions.md (hunk 3 above)
Comment (wojciechczerniak) [reply docs/guide/custom-functions.md:86]: Looks good, we will do this later.

**Review by wojciechczerniak — APPROVED** [review]
Thanks! 🏆

**Comment by izulin** [comment]
> Thanks! 🏆

should I wait for @scarletfog comments?

**Comment by wojciechczerniak** [comment]
@izulin No: https://github.com/handsontable/hyperformula/pull/569#discussion_r528697711  We have another branch waiting for this, and I'm not sure when @scarletfog will be back.

File: docs/guide/custom-functions.md (hunk 3 above)
Comment (krzysztofspilka) [reply docs/guide/custom-functions.md:86]: @wojciechczerniak @izulin I think we the proper way to write: "(...) inside the plugin, you can do this by the static `aliases` property." would be either "(...) with the static 'aliases' property" or (...) by adding/using the static 'aliases' property."

**Pull request merged**